crates/autopilot/src/decoded_settlement.rs

```diff
@@ -0,0 +1,511 @@
+}
+
+#[derive(Debug)]
+pub struct DecodedTrade {
+    pub sell_token_index: U256,
+    pub buy_token_index: U256,
+    pub receiver: Address,
+    pub sell_amount: U256,
+    pub buy_amount: U256,
+    pub valid_to: u32,
+    pub app_data: Bytes<[u8; 32]>,
+    pub fee_amount: U256,
+    pub flags: U256,
+    pub executed_amount: U256,
+    pub signature: Bytes<Vec<u8>>,
+}
+
+#[derive(Debug)]
+pub struct DecodedInteraction {
+    pub target: Address,
+    pub value: U256,
+    pub call_data: Bytes<Vec<u8>>,
+}
+
+impl From<(Address, U256, Bytes<Vec<u8>>)> for DecodedInteraction {
+    fn from((target, value, call_data): (Address, U256, Bytes<Vec<u8>>)) -> Self {
+        Self {
+            target,
+            value,
+            call_data,
+        }
+    }
+}
+
+impl From<DecodedSettlementTokenized> for DecodedSettlement {
+    fn from((tokens, clearing_prices, trades, interactions): DecodedSettlementTokenized) -> Self {
+        DecodedSettlement {
+            tokens,
+            clearing_prices,
+            trades: trades
+                .into_iter()
+                .map(
+                    |(
+                        sell_token_index,
+                        buy_token_index,
+                        receiver,
+                        sell_amount,
+                        buy_amount,
+                        valid_to,
+                        app_data,
+                        fee_amount,
+                        flags,
+                        executed_amount,
+                        signature,
+                    )| DecodedTrade {
+                        sell_token_index,
+                        buy_token_index,
+                        receiver,
+                        sell_amount,
+                        buy_amount,
+                        valid_to,
+                        app_data,
+                        fee_amount,
+                        flags,
+                        executed_amount,
+                        signature,
+                    },
+                )
+                .collect(),
+            interactions: [
+                interactions[0]
+                    .clone()
+                    .into_iter()
+                    .map(Into::into)
+                    .collect(),
+                interactions[1]
+                    .clone()
+                    .into_iter()
+                    .map(Into::into)
+                    .collect(),
+                interactions[2]
+                    .clone()
+                    .into_iter()
+                    .map(Into::into)
+                    .collect(),
+            ],
+        }
+    }
+}
+
+pub struct FeeConfiguration {
+    pub fee_objective_scaling_factor: f64,
+}
+
+impl DecodedSettlement {
+    pub fn new(contract: &GPv2Settlement, input: &[u8]) -> Result<Self> {
+        let function = contract
+            .raw_instance()
+            .abi()
+            .function("settle")
+            .context("settle function not found")?;
```

Comment:
```suggestion
    pub fn new(input: &[u8]) -> Result<Self> {
        let function = GPv2Settlement::raw_contract()
            .abi
            .function("settle")
            .unwrap();
```

No need to pass in a specific instance of the contract. Unwrap because we know at compile time that this function exists.